### PR TITLE
Validate requests before converting the network-endian to the host endian

### PR DESCRIPTION
**Pull request by nichamon**

### CODE REVIEW

File: ldms/src/ldmsd/ldmsd_request.c [hunk 1]
```diff
@@ -774,36 +773,41 @@ static int string2attr_list(char *str, struct attr_value_list **__av_list,
...
 			return 0;
 		/* Check null termination of each attr_value */
-		else if (attr->attr_value[attr->attr_len-1] != '\0')
+		else if (attr->attr_value[attr_len-1] != '\0')
 			return 0;
-		attr = (ldmsd_req_attr_t)&attr->attr_value[attr->attr_len];
-		if (attr->discrim) {
-			if (off + sizeof(*attr) + sizeof(attr->discrim) > rh->rec_len)
+		attr = (ldmsd_req_attr_t)&attr->attr_value[attr_len];
+		if (ntohl(attr->discrim)) {
```
Comment (tom95858) [inline ldms/src/ldmsd/ldmsd_request.c:809]: Byte swapping zero is a no-op

File: ldms/src/ldmsd/ldmsd_request.c [hunk 2]
```diff
@@ -774,36 +773,41 @@ static int string2attr_list(char *str, struct attr_value_list **__av_list,
...
+		attr_len = ntohl(attr->attr_len);
+
+		off += sizeof(*attr) + attr_len;
 		/* Check attr doesn't reference beyond end of request buffer */
-		if (off + sizeof(attr->discrim) > rh->rec_len)
+		if (off + sizeof(attr->discrim) > rec_len)
 			return 0;
 		/* Check null termination of each attr_value */
-		else if (attr->attr_value[attr->attr_len-1] != '\0')
+		else if (attr->attr_value[attr_len-1] != '\0')
```
Comment (tom95858) [inline ldms/src/ldmsd/ldmsd_request.c:806]: This will break if attr_len is zero. it should be something like:
```
else if (attr_len && attr->attr_value[attr_len-1] != '\0')
```

**Review by tom95858 — CHANGES_REQUESTED** [review]
Please fix that one bug with attr_len and this is good to go I think. Did you run this through the configuration regression testing? If so, we're missing tests for attributes like 'reset' which have zero length.

File: ldms/src/ldmsd/ldmsd_request.c (hunk 1 above)
Comment (nichamon) [reply ldms/src/ldmsd/ldmsd_request.c:809]: `attr->discrim` can be a non-zero value.

File: ldms/src/ldmsd/ldmsd_request.c (hunk 2 above)
Comment (nichamon) [reply ldms/src/ldmsd/ldmsd_request.c]: Thanks for catching this. I just pushed the change.

**Comment by nichamon** [comment]
I tested the change with stream and failover, and both work. When I tested stream, I included #718 .

File: ldms/src/ldmsd/ldmsd_request.c (hunk 1 above)
Comment (tom95858) [reply ldms/src/ldmsd/ldmsd_request.c:809]: You're only checking if it's zero or not. Zero is zero regardless of the byte order. If you were checking if attr->discrim equaled some value other than zero or -1 (0xff....) then you would have to byte swap it.

File: ldms/src/ldmsd/ldmsd_failover.c [hunk 3]
```diff
@@ -2355,6 +2356,7 @@ int failover_ping_handler(ldmsd_req_ctxt_t req)
 	data.state = f->state;
 	data.conn_state = f->conn_state;
 	data.flags = f->flags;
+	data.zero = 0;
```
Comment (tom95858) [inline ldms/src/ldmsd/ldmsd_failover.c:2359]: Seems like this is a change from another fix?

Comment (nichamon) [reply ldms/src/ldmsd/ldmsd_failover.c:2359]: The failover's ping responses are not null-terminated. With the validation patch, failover won't work. To fix this, @narategithub appends the responses with zero, and I included the fix to `ldmsd_failover.c` in this pull request.

**Comment by tom95858** [comment]
@nichamon can you please rebase this change?

**Comment by nichamon** [comment]
@tom95858 I just rebased the patch as requested.

**Comment by narategithub** [comment]
@nichamon I think you (or Tom) need to mark the requested change "resolved" in order for this PR to be merged. Can you also rebase this on top of OVIS-4 too?

**Review by narategithub — CHANGES_REQUESTED** [review]
ignore me .. just testing

File: ldms/src/ldmsd/ldmsd_request.c [hunk 4]
```diff
@@ -774,36 +773,41 @@ static int string2attr_list(char *str, struct attr_value_list **__av_list,
 	return rc;
 }
 
-int validate_ldmsd_req(ldmsd_req_ctxt_t reqc, ldmsd_req_hdr_t rh)
+int validate_ldmsd_req(ldmsd_req_hdr_t rh)
```
Comment (narategithub) [inline ldms/src/ldmsd/ldmsd_request.c:776]: ignore me. I'm just testing ...

**Review by narategithub — CHANGES_REQUESTED** [review]
ignore me again ... just testing.

**Review by narategithub — APPROVED** [review]

**Comment by nichamon** [comment]
@tom95858 I rebased this on top of the OVIS-4 branch. If you still cannot merge it, please approve the changes. On the 'File changed' tab, click 'review changes' and then 'approve'.
